### ISSUE TITLE
Warn on call to Map.remove with unrelated type

### ISSUE THREAD

**Issue by a14n** [dart-lang/linter#414]
```dart
final map = <String,String>{};
map.remove(1); // LINT
map.remove(""); // OK
```

**Comment by srawlins** [comment]
Seems like this can just be folded into [list_remove_unrelated_type](https://github.com/dart-lang/linter/blob/master/lib/src/rules/list_remove_unrelated_type.dart)? Can't imagine why someone would want to use one and not the other.

**Comment by a14n** [comment]
But the name of the lint is completely unrelated to map.

**Comment by srawlins** [comment]
I don't think that the name should prevent us from re-using it. We can create an alias for `list_remove_unrelated_type` called `remove_unrelated_type` (or better yet, rename the rule and create an alias for the old name).

Is there a reason to not make a breaking change? Just bump the appropriate version number.

**Comment by a14n** [comment]
>Is there a reason to not make a breaking change? Just bump the appropriate version number.

Unfortunately linter package is not like any other package. It's bundled in the sdk with the analyzer. So a simple new release of the sdk may come with a new version of linter. Actually the version number of linter is only used to [link it into the sdk](https://github.com/dart-lang/sdk/blob/f5cd6ccfb73799818ef3c88ea4b4d9e582f63504/DEPS#L98).

Note that it would be a normal package if it'd [become an anlyzer plugin](https://github.com/dart-lang/linter/issues/1097).

**Comment by bwilkerson** [comment]
> Note that it would be a normal package ...

Yes, but that comes with other costs, and it isn't clear to me that that's the best path forward for our users.

**Comment by srawlins** [comment]
Since @pq has been working on deprecation recently, what say we deprecate `list_remove_unrelated_type` and rename to `collection_remove_unrelated_type`?

**Comment by bwilkerson** [comment]
Sounds reasonable. And if we do that, then I'm going to want us to have a way to implement a quick-fix to change the name for users, which our current deprecation support doesn't include.

**Comment by munificent** [comment]
Other operations we might want to either roll into the existing lints or define new lints for:

```dart
Map.containsKey(Object? key)
Map.containsValue(Object? value)
Map.operator[Object? value]
Set.contains(Object? value)
Set.lookup(Object? object)
Set.remove(Object? value)
```

**Comment by lrhn** [comment]
The complete list (as I see it) would be:
```dart
Iterable.contains(Object?)  // inherited by List/Set/Queue too
List.remove(Object?)
Set.remove(Object?)
Set.lookup(Object?)
Set.containsAll(Iterable<Object?>)
Map.remove(Object?)
Map.containsKey(Object?)
Map.containsValue(Object?)
Map.operator [](Object?)
Queue.remove(Object?)
```
I've omitted the following:
```dart
Set.removeAll(Iterable<Object?>)
Set.retainAll(Iterable<Object?>)
Set.intersection(Set<Object?>)
Set.difference(Set<Object?>)
```
The `intersection` and `difference` makes sense for arbitrary set types, mathematically and type-wise, it's really a symmetric operation, and it's only lack of intersection types that make us choose the type of the first operand.
The `removeAll` and `retainAll`are destructive versions of the same operations, so I'd keep allowing any element type there too.
These differ from `containsAll` where having *any* member that is not a potential element of the set will guarantee a `false` result. There might still be situations where that can be useful, but I think it's, on average, better to check the iterable type to avoid mistakes.

**Comment by srawlins** [comment]
Complete with https://github.com/dart-lang/linter/pull/3692
